Composer/packages/tools/language-servers/language-generation/src/LGServer.ts

```diff
@@ -0,0 +1,355 @@
+    let text = document.getText();
+    if (LGDocument) {
+      // concat new content
+      const { content, template } = LGDocument;
+      try {
+        text = updateTemplateInContent(content, template);
+        // eslint-disable-next-line no-empty
+      } catch (error) {}
+    }
+    return text;
+  }
+
+  protected hover(params: TextDocumentPositionParams): Thenable<Hover | null> {
+    const document = this.documents.get(params.textDocument.uri);
+    if (!document) {
+      return Promise.resolve(null);
+    }
+    const text = this.getLGDocumentContent(document);
+    let templates: LGTemplate[] = [];
+    try {
+      const lgResources = getLGResources(text);
+      templates = lgResources.Templates;
+      // eslint-disable-next-line no-empty
+    } catch (error) {}
+
+    const wordRange = getRangeAtPosition(document, params.position);
+    let word = document.getText(wordRange);
+    const matchItem = templates.find(u => u.Name === word);
+    if (matchItem) {
+      const hoveritem: Hover = { contents: [matchItem.Body] };
+      return Promise.resolve(hoveritem);
+    }
+    if (word.startsWith('builtin.')) {
+      word = word.substring('builtin.'.length);
+    }
+
+    if (buildInfunctionsMap.has(word)) {
+      const functionEntity = buildInfunctionsMap.get(word);
+      if (!functionEntity) {
+        return Promise.resolve(null);
+      }
+      const hoveritem: Hover = {
+        contents: [
+          `Parameters: ${get(functionEntity, 'Params', []).join(', ')}`,
+          `Documentation: ${get(functionEntity, 'Introduction', '')}`,
+          `ReturnType: ${get(functionEntity, 'Returntype', '').valueOf()}`,
+        ],
+      };
+      return Promise.resolve(hoveritem);
+    }
+    return Promise.resolve(null);
+  }
+
+  protected async resolveSchema(url: string): Promise<string> {
+    const uri = URI.parse(url);
+    if (uri.scheme === 'file') {
+      return new Promise<string>((resolve, reject) => {
+        // eslint-disable-next-line security/detect-non-literal-fs-filename
+        fs.readFile(uri.fsPath, 'UTF-8', (err, result) => {
+          err ? reject('') : resolve(result.toString());
+        });
+      });
+    }
+    try {
+      const response = await xhr({ url, followRedirects: 5 });
+      return response.responseText;
+    } catch (error) {
+      return Promise.reject(error.responseText || getErrorStatusDescription(error.status) || error.toString());
+    }
+  }
+
+  private removeParamFormat(params: string): string {
+    const paramArr: string[] = params.split(',');
+    const resultArr: string[] = [];
+    paramArr.forEach(element => {
+      resultArr.push(element.trim().split(':')[0]);
+    });
+    return resultArr.join(' ,');
+  }
+
+  private matchedStates(params: TextDocumentPositionParams): { matched: boolean; state: string } | null {
+    const state: string[] = [];
+    const document = this.documents.get(params.textDocument.uri);
+    if (!document) return null;
```

Comment:
Return undefined is better than null here.

Undefined = I can't find. There is nothing. Nothing has been done.

Reply:
It align with ECMAScript `String.match` method, it return null. i think return null is ok.